### PR TITLE
Disallow adding ChannelHandlers with the same name to the ChannelPipeline

### PR DESCRIPTION
**Pull request by normanmaurer**
Motivation:

    We should not allow to add ChannelHandlers with the same name to the ChannelPipeline as these should be usable to uniqly identify ChannelHandlers.

    Modifications:

    - Dissallow adding ChannelHandlers with the same name to the ChannelPipeline
    - Add unit tests.

    Result:

    More robust ChannelPipeline.

### CODE REVIEW

**Comment by normanmaurer** [comment]
@swift-nio-bot test this please

**Comment by normanmaurer** [comment]
failed again because of syscall slowness... trying again

**Comment by normanmaurer** [comment]
@swift-nio-bot test this please

**Comment by normanmaurer** [comment]
@Lukasa I think we should enforce the "unique name" thing. The main reason for this is that it otherwise may result in unexpected and hard to debug problems. For example we tell users to re-use `ChannelHandler`s which may also adjust the pipeline and add other `ChannelHandler`s to it (which may have a name argument when adding it). It could happen that a user use the exact same name as the one that is used in the `ChannelHandler` that they re-use and so when trying to access their ChannelHandler or remove their `ChannelHandler` it would in fact act on the wrong one. 

Also while its true that `ChannelPipeline.add(...)` can become O(n) this is only true when the user pass a `name` when adding the `ChannelHandler` (if `nil` is used, which is the default we do not need any validation).

Even if we need validation I think its completely fine because:

- `ChannelPipeline`s usually are "short"
- `add(...)` is not an operation that I would expect to be called so frequently that it matters.

If we are super concerned about the cost we could also have an internal `Set` that just contains the names. The only downside would be more memory usage per `ChannelPipeline`.

I would also be interested to hear @weissi's take on it :)

So to summarize all of this I think we should do both changes (just as we do in Netty).

**Comment by normanmaurer** [comment]
Updated this PR to only include the handling of duplicated names...

**Comment by normanmaurer** [comment]
@Lukasa @weissi so what do we want to do with this one now ?

**Comment by Lukasa** [comment]
I continue to be in the camp of either "remove names" or "don't enforce uniqueness, encourage it with name spacing suggestions". I'm not really moved by the need to enforce uniqueness, nor do I feel the need to add a map to the pipeline in addition to a linked list.

**Comment by normanmaurer** [comment]
Also @Scott-mitch... whats your feeling about usefulness of handler naming ?

**Comment by Scottmitch** [comment]
I don't have a strong opinion but I like @Lukasa's suggestion. The primary reasons:

- not "over promise" from the perspective of the API which isn't clear if it is strictly necessary
- give implementations the freedom to enforce uniqueness or not

If channel creation is a common occurrence (relative to data manipulation after a channel is setup) then this may have larger performance impacts (depends upon the use case).

I would be very cautious about adding a `Set` to the `ChannelPipeline` for this purpose due to memory overhead and I'm not sure it is strictly necessary.

**Comment by trustin** [comment]
I'd prefer unique names because it would make an insertion of a handler at arbitrary point in the pipeline much easier. Without this constraint, there would be no point of having handler names at all.

**Comment by trustin** [comment]
Alternatively, a user could be given with some handle which denotes its location in a pipeline and he or she could use that handle to specify the insertion point. This would make handle 'names' pretty much pointless.

**Comment by Lukasa** [comment]
@trustin Isn't handler identity a sufficiently good handle? That if, if you have a ref to the handler in question you can automatically insert after or before it.

**Comment by helje5** [comment]
FWIW: I thought the name is just for debugging.

I agree with @Lukasa that identity is the right thing here (or maybe the dynamic type of you would really want to do something dynamic).

**Comment by trustin** [comment]
@Lukasa Only if we do not allow adding the same handler instance twice?

**Comment by Lukasa** [comment]
@trustin Is there any utility in adding the same handler instance twice?

**Comment by helje5** [comment]
A logger maybe? I can see a few use cases for such.
